IMPORTS

In [14]:
from pathlib import Path
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

LOCATE THE PROJECT FOLDER

In [8]:
PROJECT_ROOT = Path.cwd()

if PROJECT_ROOT.name == "notebooks":
    PROJECT_ROOT = PROJECT_ROOT.parent

print(PROJECT_ROOT)

c:\Users\hamoo\OneDrive\Desktop\PROJECTS\DATA SCIENCE\p2p-bottleneck-delinquency-predictor


DEFINE THE DATSET PATH

In [9]:
DATA_PATH = (
    PROJECT_ROOT
    / "src"
    / "data"
    / "raw"
    / "customer_invoices"
    / "dataset.csv"
)

print(DATA_PATH)
print(DATA_PATH.exists())

c:\Users\hamoo\OneDrive\Desktop\PROJECTS\DATA SCIENCE\p2p-bottleneck-delinquency-predictor\src\data\raw\customer_invoices\dataset.csv
True


LOAD THE DATASET

In [10]:
df =pd.read_csv(DATA_PATH)
print("Dataset loaded successfully")

Dataset loaded successfully


FIRST LOOK AT THE DATA

In [11]:
df.head()

,business_code,cust_number,name_customer,clear_date,buisness_year,doc_id,posting_date,document_create_date,document_create_date.1,due_in_date,invoice_currency,document type,posting_id,area_business,total_open_amount,baseline_create_date,cust_payment_terms,invoice_id,isOpen
0,U001,0200769623,WAL-MAR corp,2020-02-11 00:00:00,2020.0,1.930438e+09,2020-01-26,20200125,20200126,20200210.0,USD,RV,1.0,NaN,54273.28,20200126.0,NAH4,1.930438e+09,0
1,U001,0200980828,BEN E,2019-08-08 00:00:00,2019.0,1.929646e+09,2019-07-22,20190722,20190722,20190811.0,USD,RV,1.0,NaN,79656.60,20190722.0,NAD1,1.929646e+09,0
2,U001,0200792734,MDV/ trust,2019-12-30 00:00:00,2019.0,1.929874e+09,2019-09-14,20190914,20190914,20190929.0,USD,RV,1.0,NaN,2253.86,20190914.0,NAA8,1.929874e+09,0
3,CA02,0140105686,SYSC llc,NaN,2020.0,2.960623e+09,2020-03-30,20200330,20200330,20200410.0,CAD,RV,1.0,NaN,3299.70,20200331.0,CA10,2.960623e+09,1
4,U001,0200769623,WAL-MAR foundation,2019-11-25 00:00:00,2019.0,1.930148e+09,2019-11-13,20191113,20191113,20191128.0,USD,RV,1.0,NaN,33133.29,20191113.0,NAH4,1.930148e+09,0


In [12]:
df.tail()

,business_code,cust_number,name_customer,clear_date,buisness_year,doc_id,posting_date,document_create_date,document_create_date.1,due_in_date,invoice_currency,document type,posting_id,area_business,total_open_amount,baseline_create_date,cust_payment_terms,invoice_id,isOpen
49995,U001,0200561861,CO corporation,NaN,2020.0,1.930797e+09,2020-04-21,20200417,20200421,20200506.0,USD,RV,1.0,NaN,3187.86,20200421.0,NAA8,1.930797e+09,1
49996,U001,0200769623,WAL-MAR co,2019-09-03 00:00:00,2019.0,1.929744e+09,2019-08-15,20190814,20190815,20190830.0,USD,RV,1.0,NaN,6766.54,20190815.0,NAH4,1.929744e+09,0
49997,U001,0200772595,SAFEW associates,2020-03-05 00:00:00,2020.0,1.930537e+09,2020-02-19,20200218,20200219,20200305.0,USD,RV,1.0,NaN,6120.86,20200219.0,NAA8,1.930537e+09,0
49998,U001,0200726979,BJ'S llc,2019-12-12 00:00:00,2019.0,1.930199e+09,2019-11-27,20191126,20191127,20191212.0,USD,RV,1.0,NaN,63.48,20191127.0,NAA8,1.930199e+09,0
49999,U001,0200020431,DEC corp,2019-01-15 00:00:00,2019.0,1.928576e+09,2019-01-05,20190105,20190105,20190124.0,USD,RV,1.0,NaN,1790.30,20190101.0,NAM4,1.928576e+09,0


DATASET DIMENSIONS

In [13]:
print(f"Rows : {df.shape[0]:,}")
print(f"Columns : {df.shape[1]}")

Rows : 50,000
Columns : 19


COLUMN NAMES

In [14]:
print("Columns")
for column in df.columns:
    print(f"- {column}")

Columns
- business_code
- cust_number
- name_customer
- clear_date
- buisness_year
- doc_id
- posting_date
- document_create_date
- document_create_date.1
- due_in_date
- invoice_currency
- document type
- posting_id
- area_business
- total_open_amount
- baseline_create_date
- cust_payment_terms
- invoice_id
- isOpen


DATA TYPES

In [15]:
df.dtypes

business_code              object
cust_number                object
name_customer              object
clear_date                 object
buisness_year             float64
doc_id                    float64
posting_date               object
document_create_date        int64
document_create_date.1      int64
due_in_date               float64
invoice_currency           object
document type              object
posting_id                float64
area_business             float64
total_open_amount         float64
baseline_create_date      float64
cust_payment_terms         object
invoice_id                float64
isOpen                      int64
dtype: object

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   business_code           50000 non-null  object 
 1   cust_number             50000 non-null  object 
 2   name_customer           50000 non-null  object 
 3   clear_date              40000 non-null  object 
 4   buisness_year           50000 non-null  float64
 5   doc_id                  50000 non-null  float64
 6   posting_date            50000 non-null  object 
 7   document_create_date    50000 non-null  int64  
 8   document_create_date.1  50000 non-null  int64  
 9   due_in_date             50000 non-null  float64
 10  invoice_currency        50000 non-null  object 
 11  document type           50000 non-null  object 
 12  posting_id              50000 non-null  float64
 13  area_business           0 non-null      float64
 14  total_open_amount       50000 non-null

MISSING VALUES

In [17]:
missing_summary = (
    df.isna()
    .sum()
    .sort_values(ascending=False)
    .to_frame(name="missing_count")
)

missing_summary["missing_percentage"] = (
    (missing_summary["missing_count"] / len(df)) * 100
)

missing_summary

,missing_count,missing_percentage
area_business,50000,100.000
clear_date,10000,20.000
invoice_id,6,0.012
name_customer,0,0.000
cust_number,0,0.000
business_code,0,0.000
buisness_year,0,0.000
document_create_date,0,0.000
document_create_date.1,0,0.000
doc_id,0,0.000


DUPLICATE ROWS

In [18]:
duplicate_count = df.duplicated().sum()

print(f"Duplicate Rows: {duplicate_count:,}")

Duplicate Rows: 1,161


NUMERIC SUMMARY

In [19]:
df.describe()

,buisness_year,doc_id,document_create_date,document_create_date.1,due_in_date,posting_id,area_business,total_open_amount,baseline_create_date,invoice_id,isOpen
count,50000.000000,5.000000e+04,5.000000e+04,5.000000e+04,5.000000e+04,50000.0,0.0,50000.000000,5.000000e+04,4.999400e+04,50000.000000
mean,2019.305700,2.012238e+09,2.019351e+07,2.019354e+07,2.019368e+07,1.0,NaN,32337.021651,2.019354e+07,2.011340e+09,0.200000
std,0.460708,2.885235e+08,4.496041e+03,4.482134e+03,4.470614e+03,0.0,NaN,39205.975231,4.482701e+03,2.766335e+08,0.400004
min,2019.000000,1.928502e+09,2.018123e+07,2.018123e+07,2.018122e+07,1.0,NaN,0.720000,2.018121e+07,1.928502e+09,0.000000
25%,2019.000000,1.929342e+09,2.019050e+07,2.019051e+07,2.019052e+07,1.0,NaN,4928.312500,2.019050e+07,1.929342e+09,0.000000
50%,2019.000000,1.929964e+09,2.019091e+07,2.019091e+07,2.019093e+07,1.0,NaN,17609.010000,2.019091e+07,1.929964e+09,0.000000
75%,2020.000000,1.930619e+09,2.020013e+07,2.020013e+07,2.020022e+07,1.0,NaN,47133.635000,2.020013e+07,1.930619e+09,0.000000
max,2020.000000,9.500000e+09,2.020052e+07,2.020052e+07,2.020071e+07,1.0,NaN,668593.360000,2.020052e+07,2.960636e+09,1.000000


CATEGORICAL SUMMMARY

In [20]:
df.describe(include="object").T

,count,unique,top,freq
business_code,50000,6,U001,45359
cust_number,50000,1425,0200769623,11483
name_customer,50000,4197,WAL-MAR trust,1179
clear_date,40000,403,2019-11-12 00:00:00,309
posting_date,50000,506,2020-03-24,226
invoice_currency,50000,2,USD,46081
document type,50000,2,RV,49994
cust_payment_terms,50000,74,NAA8,20118


UNIQUE VALUE ANALYSIS

In [21]:
unique_summary = (
    df.nunique()
    .sort_values(ascending=False)
    .to_frame(name="unique_values")
)

unique_summary

,unique_values
doc_id,48839
invoice_id,48833
total_open_amount,44349
name_customer,4197
cust_number,1425
due_in_date,547
document_create_date,507
document_create_date.1,506
baseline_create_date,506
posting_date,506


INVESTIGATE

In [22]:
df["isOpen"].value_counts(dropna=False)

isOpen
0    40000
1    10000
Name: count, dtype: int64

In [23]:
df["isOpen"].value_counts(normalize=True, dropna = False) * 100

isOpen
0    80.0
1    20.0
Name: proportion, dtype: float64

INVESTIGATE THE DATE COLUMNS

In [24]:
date_candidates = [
    "clear_date",
    "posting_date",
    "document_create_date",
    "document_create_date.1",
    "due_in_date",
    "baseline_create_date"
]

for column in date_candidates:
    print(f"\n{column}")
    print(df[column].head())


clear_date
0    2020-02-11 00:00:00
1    2019-08-08 00:00:00
2    2019-12-30 00:00:00
3                    NaN
4    2019-11-25 00:00:00
Name: clear_date, dtype: object

posting_date
0    2020-01-26
1    2019-07-22
2    2019-09-14
3    2020-03-30
4    2019-11-13
Name: posting_date, dtype: object

document_create_date
0    20200125
1    20190722
2    20190914
3    20200330
4    20191113
Name: document_create_date, dtype: int64

document_create_date.1
0    20200126
1    20190722
2    20190914
3    20200330
4    20191113
Name: document_create_date.1, dtype: int64

due_in_date
0    20200210.0
1    20190811.0
2    20190929.0
3    20200410.0
4    20191128.0
Name: due_in_date, dtype: float64

baseline_create_date
0    20200126.0
1    20190722.0
2    20190914.0
3    20200331.0
4    20191113.0
Name: baseline_create_date, dtype: float64


INVESTIGATE THE SUSPICIOUS DUPLICATE DATE COLUMN

In [25]:
comparison = df["document_create_date"].astype(str) == df["document_create_date.1"].astype(str)

print(f"Matching rows: {comparison.sum():,}")
print(f"Different rows: {(~comparison).sum():,}")

Matching rows: 21,232
Different rows: 28,768


# Initial Findings

## Dataset Overview

- Rows: 50,000
- Columns: 19
- Duplicate rows: 1,161

## Potential Target

- `isOpen`: 

## Potential Identifier Columns

- `cust_number`:
- `doc_id`:
- `invoice_id`:

## Date Columns

- `clear_date`:
- `posting_date`:
- `document_create_date`:
- `document_create_date.1`:
- `due_in_date`:
- `baseline_create_date`:

## Data Quality Observations

- Missing values:
- Duplicate records:
- Unexpected data types:
- Other observations:

## Questions Requiring Further Investigation

1.
2.
3.

INVESTIGATION 1 - Is Open Exactly tied to clear_date?

In [26]:
pd.crosstab(
    df["isOpen"],
    df["clear_date"].isna()
)

clear_date,False,True
isOpen,,
0,40000,0
1,0,10000


INVESTIGATION 2 - See actual date values

In [27]:
print(df["posting_date"].head(10).to_list())
print(df["clear_date"].head(10).to_list())
print(df["due_in_date"].head(10).to_list())
print(df["baseline_create_date"].head(10).to_list())
print(df["document_create_date"].head(10).to_list())
print(df["document_create_date.1"].head(10).to_list())

['2020-01-26', '2019-07-22', '2019-09-14', '2020-03-30', '2019-11-13', '2019-09-20', '2019-11-01', '2020-03-19', '2019-06-07', '2019-02-20']
['2020-02-11 00:00:00', '2019-08-08 00:00:00', '2019-12-30 00:00:00', nan, '2019-11-25 00:00:00', '2019-12-04 00:00:00', '2019-11-12 00:00:00', nan, '2019-06-18 00:00:00', '2019-03-06 00:00:00']
[20200210.0, 20190811.0, 20190929.0, 20200410.0, 20191128.0, 20191004.0, 20191116.0, 20200403.0, 20190622.0, 20190307.0]
[20200126.0, 20190722.0, 20190914.0, 20200331.0, 20191113.0, 20190924.0, 20191101.0, 20200319.0, 20190607.0, 20190220.0]
[20200125, 20190722, 20190914, 20200330, 20191113, 20190920, 20191031, 20200318, 20190605, 20190219]
[20200126, 20190722, 20190914, 20200330, 20191113, 20190920, 20191101, 20200319, 20190607, 20190220]


INVESTIGATION 3 - Understand the payment terms

In [28]:
df["cust_payment_terms"].value_counts().head(20)

cust_payment_terms
NAA8    20118
NAH4    13585
CA10     3800
NAC6     1743
NAM4     1385
NAAX     1208
NAD1      927
NAU5      912
NAGD      906
NAG2      808
NA32      690
NAX2      620
NAM2      600
NA10      434
NAD5      318
NAVE      260
NAM1      257
NAVF      217
NAAW      192
NAWU      117
Name: count, dtype: int64

Investigation 4 - Understand document type

In [29]:
df["document type"].value_counts(dropna=False)

document type
RV    49994
X2        6
Name: count, dtype: int64

In [30]:
df["invoice_currency"].value_counts(dropna=False)

invoice_currency
USD    46081
CAD     3919
Name: count, dtype: int64

INVESTIGATION 5 - Total Open amount breakdown

In [31]:
df.groupby("isOpen")["total_open_amount"].agg(
    ["count", "mean", "median", "min", "max"]
)

,count,mean,median,min,max
isOpen,,,,,
0,40000,32065.034894,17283.5,0.79,668593.36
1,10000,33424.968682,19344.8,0.72,653644.80


DATE RELATIONSHIPS

In [32]:
date_columns = [
    "posting_date",
    "document_create_date",
    "document_create_date.1",
    "due_in_date",
    "baseline_create_date",
    "clear_date",
]

for column in date_columns:
    print(f"\n{column}")
    print(f"Min: {df[column].min()}")
    print(f"Max: {df[column].max()}")


posting_date
Min: 2018-12-30
Max: 2020-05-22

document_create_date
Min: 20181226
Max: 20200522

document_create_date.1
Min: 20181230
Max: 20200522

due_in_date
Min: 20181224.0
Max: 20200710.0

baseline_create_date
Min: 20181214.0
Max: 20200522.0

clear_date


TypeError: '<=' not supported between instances of 'str' and 'float'

In [33]:
df[
    [
        "posting_date",
        "document_create_date",
        "document_create_date.1",
        "baseline_create_date",
        "due_in_date",
        "clear_date",
        "cust_payment_terms",
    ]
].head(20)

,posting_date,document_create_date,document_create_date.1,baseline_create_date,due_in_date,clear_date,cust_payment_terms
0,2020-01-26,20200125,20200126,20200126.0,20200210.0,2020-02-11 00:00:00,NAH4
1,2019-07-22,20190722,20190722,20190722.0,20190811.0,2019-08-08 00:00:00,NAD1
2,2019-09-14,20190914,20190914,20190914.0,20190929.0,2019-12-30 00:00:00,NAA8
3,2020-03-30,20200330,20200330,20200331.0,20200410.0,NaN,CA10
4,2019-11-13,20191113,20191113,20191113.0,20191128.0,2019-11-25 00:00:00,NAH4
5,2019-09-20,20190920,20190920,20190924.0,20191004.0,2019-12-04 00:00:00,CA10
6,2019-11-01,20191031,20191101,20191101.0,20191116.0,2019-11-12 00:00:00,NAH4
7,2020-03-19,20200318,20200319,20200319.0,20200403.0,NaN,NAA8
8,2019-06-07,20190605,20190607,20190607.0,20190622.0,2019-06-18 00:00:00,NAH4
9,2019-02-20,20190219,20190220,20190220.0,20190307.0,2019-03-06 00:00:00,NAC6


DAYS LATE VALUES CHECK

In [34]:
analysis_df = df.copy()

analysis_df["clear_date_parsed"] = pd.to_datetime(
    analysis_df["clear_date"],
    errors="coerce"
)

analysis_df["due_in_date_parsed"] = pd.to_datetime(
    analysis_df["due_in_date"].astype("int64").astype(str),
    format="%Y%m%d",
    errors="coerce"  
)

analysis_df["days_late"] = (
    analysis_df["clear_date_parsed"] 
    -
    analysis_df["due_in_date_parsed"]
).dt.days

In [35]:
analysis_df["days_late"].describe()

count    40000.000000
mean         0.837700
std         10.831701
min        -89.000000
25%         -3.000000
50%          0.000000
75%          2.000000
max        204.000000
Name: days_late, dtype: float64

In [36]:
analysis_df["days_late"].isna().sum()

np.int64(10000)

In [37]:
analysis_df["days_late"].sort_values().head(20)

6753    -89.0
26356   -87.0
49212   -79.0
34359   -75.0
31578   -75.0
32970   -75.0
20467   -75.0
39648   -75.0
44395   -74.0
18109   -74.0
42037   -74.0
12047   -74.0
46441   -74.0
2317    -72.0
18272   -71.0
35355   -69.0
14811   -65.0
16393   -64.0
3206    -64.0
47874   -64.0
Name: days_late, dtype: float64

In [38]:
analysis_df["days_late"].sort_values(ascending=False).head(20)

20393    204.0
16990    202.0
5561     195.0
22212    171.0
42123    157.0
41558    151.0
24112    123.0
23316    122.0
8159     122.0
41586    122.0
46512    121.0
26919    121.0
34812    121.0
3131     120.0
20248    120.0
25377    120.0
34035    120.0
10023    120.0
10268    120.0
49238    120.0
Name: days_late, dtype: float64

DISTRIBUTION OF LATE VS ON-TIME

In [39]:
analysis_df["is_delayed"] = (
    analysis_df["days_late"] > 0
).astype("int64")

In [40]:
analysis_df["is_delayed"].value_counts(dropna=False)

is_delayed
0    33236
1    16764
Name: count, dtype: int64

In [41]:
analysis_df["is_delayed"].value_counts(
    normalize=True,
    dropna=False
) * 100

is_delayed
0    66.472
1    33.528
Name: proportion, dtype: float64

CHECK: NEGATIVE DAYS

In [42]:
analysis_df.loc[
    analysis_df["days_late"] < 0,
    [
        "posting_date",
        "due_in_date",
        "clear_date",
        "days_late",
        "cust_payment_terms"
    ]
].head(20)

,posting_date,due_in_date,clear_date,days_late,cust_payment_terms
1,2019-07-22,20190811.0,2019-08-08 00:00:00,-3.0,NAD1
4,2019-11-13,20191128.0,2019-11-25 00:00:00,-3.0,NAH4
6,2019-11-01,20191116.0,2019-11-12 00:00:00,-4.0,NAH4
8,2019-06-07,20190622.0,2019-06-18 00:00:00,-4.0,NAH4
9,2019-02-20,20190307.0,2019-03-06 00:00:00,-1.0,NAC6
18,2019-11-15,20191130.0,2019-11-26 00:00:00,-4.0,NAH4
19,2020-01-24,20200208.0,2020-02-05 00:00:00,-3.0,NAH4
21,2019-08-21,20190905.0,2019-09-04 00:00:00,-1.0,NAA8
23,2019-07-17,20190801.0,2019-07-30 00:00:00,-2.0,NAA8
24,2020-01-24,20200208.0,2020-02-04 00:00:00,-4.0,NAH4


RELATIONSHIP: Document Creation Date vs Posting Date

In [43]:
analysis_df["posting_date_parsed"] = pd.to_datetime(
    analysis_df["posting_date"], errors="coerce"
)

analysis_df["document_create_date_parsed"] = pd.to_datetime(
    analysis_df["document_create_date"].astype(str),
    format="%Y%m%d",
    errors="coerce"
)

analysis_df["document_create_date.1_parsed"] = pd.to_datetime(
    analysis_df["document_create_date.1"].astype(str),
    format="%Y%m%d",
    errors="coerce"
)

In [44]:
print(
    "document_create_date == posting_date:",
    (
        analysis_df["document_create_date_parsed"]
        == analysis_df["posting_date_parsed"]
    ).mean()
)

print(
    "document_create_date.1 == posting_date:",
    (
        analysis_df["document_create_date.1_parsed"]
        == analysis_df["posting_date_parsed"]
    ).mean()
)

document_create_date == posting_date: 0.42476
document_create_date.1 == posting_date: 0.99988


# Data Understanding Conclusion

The initial dataset contains 50,000 invoice records and 19 columns.

The `isOpen` field was determined not to be a suitable ML target because it is directly associated with the presence or absence of `clear_date`.

For the 40,000 cleared invoices, payment behavior can be measured using the difference between `clear_date` and `due_in_date`.

The following target definitions have been established:

- `days_late = clear_date - due_in_date`
- `is_delayed = 1` when `days_late > 0`
- `is_delayed = 0` when `days_late <= 0`

The historical data contains 33.528% delayed invoices and 66.472% invoices that were paid on or before the due date.

Negative `days_late` values represent early payments and are considered valid business outcomes.

The `document_create_date.1` column matches `posting_date` for 99.988% of records and will be investigated during data cleaning as a potential duplicate/extraction artifact.

The dataset contains 10,000 currently open invoices for which the final payment outcome is not yet known. These records will not be used to calculate historical delinquency outcomes but may eventually be used as the population for production prediction.

The next phase will focus on data validation, standardizing date fields, handling missing values, identifying redundant columns, and preparing a clean dataset for target construction and feature engineering.

In [46]:
analysis_df.head(2)

,business_code,cust_number,name_customer,clear_date,buisness_year,doc_id,posting_date,document_create_date,document_create_date.1,due_in_date,...,cust_payment_terms,invoice_id,isOpen,clear_date_parsed,due_in_date_parsed,days_late,is_delayed,posting_date_parsed,document_create_date_parsed,document_create_date.1_parsed
0,U001,0200769623,WAL-MAR corp,2020-02-11 00:00:00,2020.0,1.930438e+09,2020-01-26,20200125,20200126,20200210.0,...,NAH4,1.930438e+09,0,2020-02-11,2020-02-10,1.0,1,2020-01-26,2020-01-25,2020-01-26
1,U001,0200980828,BEN E,2019-08-08 00:00:00,2019.0,1.929646e+09,2019-07-22,20190722,20190722,20190811.0,...,NAD1,1.929646e+09,0,2019-08-08,2019-08-11,-3.0,0,2019-07-22,2019-07-22,2019-07-22


In [2]:
from pathlib import Path
import sys

PROJECT_ROOT = Path.cwd().parent
sys.path.append(str(PROJECT_ROOT))

In [8]:
from src.data.cleaner import (
    standardize_column_names,
    validate_invoice_data,
    remove_unusable_columns,
    standardize_date_columns,
    validate_date_quality,
    compare_document_and_posting_dates,
    investigate_secondary_document_dates,
    remove_redundant_columns,
    investigate_invoice_identifiers,
)
from src.data.loader import load_invoice_data

In [9]:
PROJECT_ROOTS = Path.cwd()

if PROJECT_ROOTS.name == "notebooks":
    PROJECT_ROOTS = PROJECT_ROOTS.parent

print(PROJECT_ROOTS)

c:\Users\hamoo\OneDrive\Desktop\PROJECTS\DATA SCIENCE\p2p-bottleneck-delinquency-predictor


In [10]:
DATA_PATH = (
    PROJECT_ROOTS
    / "src"
    / "data"
    / "raw"
    / "customer_invoices"
    / "dataset.csv"
)

In [11]:
df = load_invoice_data(DATA_PATH)
cleaned_df = standardize_column_names(df)
cleaned_df = remove_unusable_columns(cleaned_df)
cleaned_df = standardize_date_columns(cleaned_df)
cleaned_df = remove_redundant_columns(cleaned_df)

In [12]:
cleaned_df.head(2)

,business_code,cust_number,name_customer,clear_date,business_year,doc_id,posting_date,document_create_date,due_in_date,invoice_currency,document_type,total_open_amount,baseline_create_date,cust_payment_terms,invoice_id,isOpen
0,U001,0200769623,WAL-MAR corp,2020-02-11,2020.0,1.930438e+09,2020-01-26,2020-01-25,2020-02-10,USD,RV,54273.28,2020-01-26,NAH4,1.930438e+09,0
1,U001,0200980828,BEN E,2019-08-08,2019.0,1.929646e+09,2019-07-22,2019-07-22,2019-08-11,USD,RV,79656.60,2019-07-22,NAD1,1.929646e+09,0


In [17]:
def investigate_duplicate_records(
        df: pd.DataFrame
) -> None:
    """
    Investigate exact duplicate records and identify 
    whether duplicate invoice records contain differences
    """

    print("\n--- Duplicate Record Investigation ---")

    duplicate_rows = df[
        df.duplicated(
            keep=False
        )
    ].copy()

    print("Exact duplicated rows:", len(duplicate_rows))

    print("Duplicated groups:",duplicate_rows.duplicated().sum())

    # Investigate duplicates based on Invoice IDs
    if "invoice_id" in df.columns:
        invoice_duplicates = df[
            df["invoice_id"].duplicated(
                keep=False
            )
            & df["invoice_id"].notna()
        ].copy()

        print("\nRows with duplicated invoice_id:", len(invoice_duplicates))

        # Remove invoice_id and doc_id from comparison
        # to determine whether the remaining business
        # attributes are identical.

        comparison_columns = [
            column for column in invoice_duplicates.columns
            if column not in [
                "invoice_id",
                "doc_id",

            ]
        ] 

        business_duplicates = (
            invoice_duplicates
            .duplicated(
                subset = comparison_columns,
                keep = False,
            )
        )

        print("Duplicated invoice records with identical business attributes:", business_duplicates.sum(),)

        print("Duplicated invoice records with identical business attributes:", (~business_duplicates).sum(),)

In [18]:
investigate_duplicate_records(cleaned_df)


--- Duplicate Record Investigation ---
Exact duplicated rows: 2308
Duplicated groups: 1161

Rows with duplicated invoice_id: 2308
Duplicated invoice records with identical business attributes: 2308
Duplicated invoice records with identical business attributes: 0


In [19]:
cleaned_df_nb = cleaned_df.copy()

In [21]:
duplicated_rows = cleaned_df_nb.duplicated().sum()
print("Duplicated rows:", duplicated_rows)

deduplicated_df  = cleaned_df_nb.drop_duplicates()
print("Original shape:", cleaned_df_nb.shape)
print("After deduplication:", deduplicated_df.shape)

Duplicated rows: 1161
Original shape: (50000, 16)
After deduplication: (48839, 16)


IDENTIFIER CLEANUP

In [22]:
print("invoice_id missing:", deduplicated_df["invoice_id"].isna().sum())
print("doc_id missing:", deduplicated_df["doc_id"].isna().sum())

print(
    "invoice_id==doc_id:",
    (deduplicated_df["invoice_id"] == deduplicated_df["doc_id"]).sum()
)

print(
    "invoice_id != doc_id:",
    (deduplicated_df["invoice_id"] != deduplicated_df["doc_id"]).sum()
)

invoice_id missing: 6
doc_id missing: 0
invoice_id==doc_id: 48833
invoice_id != doc_id: 6


In [23]:
deduplicated_df[
    deduplicated_df["invoice_id"].isna()
][["invoice_id", "doc_id"]]

,invoice_id,doc_id
5095,NaN,9.500000e+09
19450,NaN,9.500000e+09
30322,NaN,9.500000e+09
31009,NaN,9.500000e+09
38443,NaN,9.500000e+09
44500,NaN,9.500000e+09


In [24]:
deduplicated_df[
    deduplicated_df["doc_id"].isna()
][["invoice_id", "doc_id"]]

,invoice_id,doc_id


MOVING UNUSED/UNNECSSARY FUNCTIONS FROM CLEANER TO NOTEBOOK FOR ARCHIVING

In [ ]:
def remove_redundant_columns(
        df: pd.DataFrame,
        columns_to_remove: List[str],
) -> pd.DataFrame:
    """
    Remove columns that are redundant or not required 
    for downstream modeling
    """

    cleaned_df = df.copy()

    # columns_to_remove = [
    #     "document_create_date_secondary"
    # ]

    cleaned_df = cleaned_df.drop(
        columns=columns_to_remove,
        errors="ignore",
    )

    return cleaned_df